clar2wasm/src/duck_type.rs

```diff
@@ -0,0 +1,581 @@
+                let target_locs = self.create_locals_for_ty(target_elem_ty);
+
+                // iterate through elements, convert them to target type and store them
+                let loop_id = {
+                    let mut loop_ = builder.dangling_instr_seq(None);
+                    let loop_id = loop_.id();
+
+                    let og_elem_size = self.read_from_memory(&mut loop_, *offset, 0, og_elem_ty)?;
+                    self.duck_type_stack(&mut loop_, og_elem_ty, target_elem_ty, &target_locs)?;
+                    for l in target_locs.iter().rev() {
+                        loop_.local_get(*l);
+                    }
+                    let target_elem_size =
+                        self.write_to_memory(&mut loop_, offset_target, 0, target_elem_ty)?;
+
+                    loop_
+                        .local_get(*offset)
+                        .i32_const(og_elem_size)
+                        .binop(BinaryOp::I32Add)
+                        .local_set(*offset);
+                    loop_
+                        .local_get(offset_target)
+                        .i32_const(target_elem_size as i32)
+                        .binop(BinaryOp::I32Add)
+                        .local_set(offset_target);
+                    loop_
+                        .local_get(*length)
+                        .i32_const(1)
+                        .binop(BinaryOp::I32Sub)
+                        .local_tee(*length)
+                        .br_if(loop_id);
+
+                    loop_id
+                };
+
+                // we will "duck-type-clone" if the length of the list is not empty
+                builder.local_get(*length).if_else(
+                    None,
+                    |then| {
+                        // we set the offset_target to copy at the free space of stack-pointer and we move this on further
+                        then.global_get(self.stack_pointer)
+                            .local_tee(offset_target)
+                            .i32_const(get_type_in_memory_size(target_ty, false))
+                            .binop(BinaryOp::I32Add)
+                            .global_set(self.stack_pointer);
+
+                        // we put the resulting offset/length on the stack
+                        then.local_get(offset_target).local_get(*length);
+
+                        // the cloning loop
+                        then.instr(Loop { seq: loop_id });
+
+                        // we set the result back to the correct locals
+                        then.local_set(*length).local_set(*offset);
+                    },
+                    |_else| {},
+                );
+            }
+            (TypeSignature::ListUnionType(_), TypeSignature::ListUnionType(_)) => {
+                return Err(GeneratorError::InternalError(
+                    "Unconcretized ListUnionType".to_owned(),
+                ))
+            }
+            _ => todo!(),
```

Comment:
Can we return an appropriate `GeneratorError` here to avoid `panic` if this line is somehow reached.

Reply:
Oof, can't believe I missed that. Thanks!

Reply:
fixed in 256afcee41134f9c7a951474b0da987d4f03e962